sw/device/tests/usbdev_mem_test.c

```diff
@@ -0,0 +1,212 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// USB packet memory test
+//
+// This test performs a simple read/write test of the entire USB packet memory
+// from the CPU to check connectivity, addressing, integrity etc. A LFSR-
+// generated byte stream is used as the test pattern.
+//
+// Two methods of access are tested in turn:
+// (i)  direct access using the mmio_ routines
+// (ii) DIF-based buffer reading/writing
+
+#include "sw/device/lib/dif/dif_pinmux.h"
+#include "sw/device/lib/dif/dif_usbdev.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/lib/testing/pinmux_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"  // Generated.
+#include "sw/device/lib/dif/autogen/dif_usbdev_autogen.h"
+
+#define USBDEV_BASE_ADDR TOP_EARLGREY_USBDEV_BASE_ADDR
+
+//#include "sw/device/lib/dif/usbdev_regs.h"  // Generated.
+#ifndef USBDEV_BUFFER_REG_OFFSET
+#define USBDEV_BUFFER_REG_OFFSET 0x800u
+#endif
+
+// Simple LFSR for 8-bit sequences
+#define LFSR_ADVANCE(lfsr) \
+  (((lfsr) << 1) ^         \
+   ((((lfsr) >> 1) ^ ((lfsr) >> 2) ^ ((lfsr) >> 3) ^ ((lfsr) >> 7)) & 1U))
+
+// Total size of packet buffer memory in bytes
+#define USBDEV_PACKET_MEM_SIZE (USBDEV_NUM_BUFFERS * USBDEV_MAX_PACKET_SIZE)
+
+// Verbose repoorting?
+// Note: this will make a massive difference to the run time on t-l simulation
+static const bool verbose = false;
+
+// Buffer descriptions
+static dif_usbdev_buffer_t buf[USBDEV_NUM_BUFFERS];
+
+// Basic test of packet memory function from the CPU side
+static uint8_t testpatt[USBDEV_PACKET_MEM_SIZE];
+static uint8_t testcopy[USBDEV_PACKET_MEM_SIZE];
+
+static dif_usbdev_t usbdev;
+static dif_usbdev_buffer_pool_t buffer_pool;
+
+OTTF_DEFINE_TEST_CONFIG();
+
+/* The obvious advantage of using the DIF interface is that we are testing that
+   code too, but the disadvantage is that we have to resort to modifying the
+   buffer descriptors to perform reads because the buffers are not being used
+   in accordance with the expected flow model.
```

Comment:
Like any guidelines, there are always exceptions :). That said, there are two things to point out here:
1. DIFs themselves are test code, so there is no need to test the test code hehe,
2. since DIFs are aimed at aiding the writing of tests, if the existing API isn't useful for such purposes I think it makes sense to change the DIFs to fit the tests needs.

As I pointed out [here](https://github.com/lowRISC/opentitan/pull/17436#discussion_r1145346152): there are not too many tests / example programs using the USB DIFs at them moment so it should be non-invasive change.

All of that said, I think @a-will should weigh in. He has more context around the OT usbdev stack (HW and SW), and was the latest author of the usbdev DIFs I believe.

Reply:
For me, either way is fine. The DIFs are generally oriented towards use cases for the IP, but it probably makes sense to add the raw read/write functions to the DIFs.

I'll point out that the test relies on functionality that usbdev isn't really meant to guarantee (that you can treat its buffer storage like an ordinary read/write memory)--It just happens to be true. Ordinarily, we'd leave this check to simulation, where we can read and write it through an HDL back door.